### PR TITLE
merge to main: Switching from Spring to Spring Boot 🚀

### PR DESCRIPTION
**Pull request by Atabek04**

### CODE REVIEW

File: src/main/java/com/epam/gym/repository/UserRepository.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+    @Modifying
+    @Query("UPDATE User u SET u.password = :newPassword WHERE u.username = :username")
+    void changePassword(String username, String newPassword);
+
+    @Modifying
+    @Query("UPDATE User u SET u.isActive = true WHERE u.username = :username AND u.isActive = false")
+    void activateUser(String username);
+
+    @Modifying
```
Comment (m-nurbek) [inline src/main/java/com/epam/gym/repository/UserRepository.java:24]: also here

File: src/main/java/com/epam/gym/service/impl/TraineeServiceImpl.java [hunk 2]
```diff
@@ -35,25 +35,29 @@
...
-    private final TrainerDAO trainerDAO;
-    private final TrainingDAO trainingDAO;
+    private final TrainerRepository trainerRepository;
+    private final TrainingRepository trainingRepository;
 
     @Transactional
     @Override
     public UserCredentials create(TraineeRequest request) {
         log.info("Creating trainee");
         var createdUser = userService.create(toUser(request))
```
Comment (m-nurbek) [inline src/main/java/com/epam/gym/service/impl/TraineeServiceImpl.java:55]: Where 'toUser' method came from? Consider importing just a Utility class, not concrete methods

```java
UserMapper.toUser(request)
```

File: src/test/java/com/epam/gym/api/TraineeControllerIntegrationTest.java [hunk 3]
```diff
@@ -49,9 +49,9 @@
 import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
 import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.put;
 import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
-import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
 import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
 
+@ActiveProfiles("test")
 @ExtendWith({SpringExtension.class, TraineeServiceParameterResolver.class})
```
Comment (m-nurbek) [inline src/test/java/com/epam/gym/api/TraineeControllerIntegrationTest.java:55]: Why just not annotate it with SpringBootTest instead of your annotations?

File: src/test/java/com/epam/gym/api/TraineeControllerIntegrationTest.java
Comment (m-nurbek) [inline src/test/java/com/epam/gym/api/TraineeControllerIntegrationTest.java:1]:
```java
@BeforeEach
    public void setup() {
        MockitoAnnotations.openMocks(this);
        mockMvc = MockMvcBuilders.standaloneSetup(traineeController)
                .setControllerAdvice(new GlobalExceptionHandler()) // Add global exception handler if needed
                .addFilter(new CharacterEncodingFilter("UTF-8", true)) // Handle character encoding
                .build();
    }
```

You have some useless comments here

Comment (m-nurbek) [inline src/test/java/com/epam/gym/api/TraineeControllerIntegrationTest.java:1]:
```java
@BeforeEach
    public void setup() {
        MockitoAnnotations.openMocks(this);
        mockMvc = MockMvcBuilders.standaloneSetup(traineeController)
                .setControllerAdvice(new GlobalExceptionHandler()) // Add global exception handler if needed
                .addFilter(new CharacterEncodingFilter("UTF-8", true)) // Handle character encoding
                .build();
    }
```

You have some useless comments here

File: src/test/java/com/epam/gym/service/TrainingServiceTest.java [hunk 4]
```diff
@@ -39,60 +40,60 @@ void testUpdateTrainingSuccess() {
         Long trainingId = 1L;
```
Comment (m-nurbek) [inline src/test/java/com/epam/gym/service/TrainingServiceTest.java:40]: Its not super necessary but I would like to see the comments like //given, //when, //then. This will help me to see what is given, what is tested, and what is asserted.
Consider adding them to your tests

Comment (m-nurbek) [inline src/test/java/com/epam/gym/service/TrainingServiceTest.java:40]: Its not super necessary but I would like to see the comments like //given, //when, //then. This will help me to see what is given, what is tested, and what is asserted.
Consider adding them to your tests

File: src/test/java/com/epam/gym/api/TrainerControllerIntegrationTest.java
Comment (m-nurbek) [inline src/test/java/com/epam/gym/api/TrainerControllerIntegrationTest.java:1]: If this is an IntegrationTest then why are you mocking something?

**Review by m-nurbek — CHANGES_REQUESTED** [review]
Good job!

Please review my comments.
I will continue to add more comments about your project, since I can see there is a job to be done to improve it. My suggestion are to improve your code style, remove unnecessary configurations in your application and tests. I also have questions about your integration tests that mock something but shouldn't do so. 
Also, I noticed that you only use one application.yml file to configure your spring boot project and I don't see any profiles inside it. You should change your config for different profiles, this can easily be done with just naming your application.yml file properly (e.g. application-dev.yml, application-prod.yml).

Best regards,
Nurbek

File: src/main/java/com/epam/gym/actuator/ExternalServiceHealthIndicator.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+        boolean serviceUp = checkExternalServiceHealth();
+
+        if (serviceUp) {
+            return Health.up().withDetail("External Service", "Available").build();
+        } else {
+            return Health.down().withDetail("External Service", "Not Available").build();
+        }
+    }
+
+    private boolean checkExternalServiceHealth() {
```
[the remark this answers is not in the document]
Comment (Atabek04) [reply src/main/java/com/epam/gym/actuator/ExternalServiceHealthIndicator.java:21]: I just left that as a specification or todo for future. I will implement it in Security Model.

File: src/main/java/com/epam/gym/controller/AuthController.java [hunk 6]
```diff
@@ -21,7 +21,7 @@
 public class AuthController {
     private final UserService userService;
 
-    @GetMapping("/login")
+    @PostMapping("/login")
     public ResponseEntity<String> login(@Valid @RequestBody UserCredentials credentials) {
```
[the remark this answers is not in the document]
Comment (Atabek04) [reply src/main/java/com/epam/gym/controller/AuthController.java:25]: I prefer leaving my old version of code for now. Cause as I understand it just removes `return ResponseEntity` part, and make your code look cleaner. Surely I will use that best practice in my future controllers.

File: src/main/java/com/epam/gym/service/impl/TraineeServiceImpl.java (hunk 2 above)
Comment (Atabek04) [reply src/main/java/com/epam/gym/service/impl/TraineeServiceImpl.java:55]: I understand that there might be a little bit context loss. But static import makes my code more readable and help to avoid multiple repetition, cause I used `toUser()` a lot. But I got you.

I will consider using fully qualified method call, when the context isn't obvious.

File: src/test/java/com/epam/gym/api/TraineeControllerIntegrationTest.java (hunk 3 above)
Comment (Atabek04) [reply src/test/java/com/epam/gym/api/TraineeControllerIntegrationTest.java:55]: Cause if I'm not mistaken `@SpringBootTest` is used to load the full application context. And in your application you should choose which profile to use (for example 'test'). But in my case I have separate test profile in test package, that's why I'm using `@ActiveProfiles("test")` at the same time.

I will fix it when I'll add rest of the profiles

File: src/test/java/com/epam/gym/api/TrainerControllerIntegrationTest.java
Comment (Atabek04) [reply src/test/java/com/epam/gym/api/TrainerControllerIntegrationTest.java:1]: yes, you're right. I totally forget about that.

File: pom.xml [hunk 7]
```diff
@@ -238,35 +172,55 @@
...
         <plugins>
+            <plugin>
+                <groupId>org.springframework.boot</groupId>
+                <artifactId>spring-boot-maven-plugin</artifactId>
+            </plugin>
+
             <plugin>
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-compiler-plugin</artifactId>
                 <version>3.10.1</version>
```
Comment (maximelyashevich) [inline pom.xml:208]: Please extract to Maven properties.

File: pom.xml [hunk 8]
```diff
@@ -238,35 +172,55 @@
...
-                        <arg>-parameters</arg>
-                    </compilerArgs>
                 </configuration>
             </plugin>
 
             <plugin>
-                <artifactId>maven-war-plugin</artifactId>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-surefire-plugin</artifactId>
+                <version>3.2.5</version>
```
Comment (maximelyashevich) [inline pom.xml:218]: Please extract to Maven properties.

File: src/main/java/com/epam/gym/GymApp.java [hunk 9]
```diff
@@ -0,0 +1,13 @@
+package com.epam.gym;
+
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.web.servlet.config.annotation.EnableWebMvc;
+
+@SpringBootApplication
+@EnableWebMvc
```
Comment (maximelyashevich) [inline src/main/java/com/epam/gym/GymApp.java:8]: What is the reason to use this annotation? Won't `@SpringBootApplication` cover web-related logic?

File: src/main/java/com/epam/gym/actuator/DiskSpaceHealthIndicator.java [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+import org.springframework.boot.actuate.health.Health;
+import org.springframework.boot.actuate.health.HealthIndicator;
+import org.springframework.stereotype.Component;
+
+import java.io.File;
+
+@Component
+public class DiskSpaceHealthIndicator implements HealthIndicator {
+
+    private static final long THRESHOLD = 100 * 1024 * 1024;  // 100 MB threshold
```
Comment (maximelyashevich) [inline src/main/java/com/epam/gym/actuator/DiskSpaceHealthIndicator.java:12]:
```suggestion
    private static final long THRESHOLD = 100L * 1024L * 1024L;  // 100 MB threshold
```

File: src/main/java/com/epam/gym/actuator/DiskSpaceHealthIndicator.java [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+    @Override
+    public Health health() {
+        File disk = new File("/");
+        long freeSpace = disk.getFreeSpace();
+
+        if (freeSpace >= THRESHOLD) {
+            return Health.up().withDetail("Free Disk Space", freeSpace).build();
+        } else {
+            return Health.down().withDetail("Free Disk Space", freeSpace).build();
+        }
```
Comment (maximelyashevich) [inline src/main/java/com/epam/gym/actuator/DiskSpaceHealthIndicator.java:23]: Duplicated String, please extract to the constant.

File: src/main/java/com/epam/gym/actuator/ExternalServiceHealthIndicator.java [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+
+    @Override
+    public Health health() {
+        boolean serviceUp = checkExternalServiceHealth();
+
+        if (serviceUp) {
+            return Health.up().withDetail("External Service", "Available").build();
+        } else {
+            return Health.down().withDetail("External Service", "Not Available").build();
+        }
```
Comment (maximelyashevich) [inline src/main/java/com/epam/gym/actuator/ExternalServiceHealthIndicator.java:18]: Let's rename `serviceUp` to `isServiceUp` to increade code readability:
```suggestion
        boolean isServiceUp = checkExternalServiceHealth();

        if (isServiceUp) {
            return Health.up().withDetail("External Service", "Available").build();
        } else {
            return Health.down().withDetail("External Service", "Not Available").build();
        }
```

Please introduce String constants for duplicated String literals.

File: src/main/java/com/epam/gym/controller/AuthController.java [hunk 13]
```diff
@@ -21,7 +21,7 @@
...
         log.info("Login attempt for user: {}", credentials.username());
```
Comment (maximelyashevich) [inline src/main/java/com/epam/gym/controller/AuthController.java:26]: I think debug level is more appropriate here:
```suggestion
        log.debug("Login attempt for user: {}", credentials.username());
```

File: src/main/java/com/epam/gym/controller/AuthController.java (hunk 6 above)
Comment (maximelyashevich) [inline src/main/java/com/epam/gym/controller/AuthController.java:25]: I suggest considering:
1. `@ResponseStatus()` annotation to declaratively define status code.
2. Eliminate the use of `ResponseEntity` and keep only your wrapped type.
3. Delegate business logic to service layer like throwing an exception in case of unauthenticated action that later will be processed by your exception handler.
4. Please be aware that HTTP 200 OK is the default status code, thus, no need to declare it explicitly.

After code refactoring, your controller method may look like this:
```Java
@PostMapping("/login")
public void login(@Valid @RequestBody UserCredentials credentials) {
     userService.tryToAuthenticate(credentials);
}
```

Please apply the similar refactoring for all your controllers.

File: src/main/java/com/epam/gym/util/UpdateEntityFields.java [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    public static void updateTrainingFields(Training existingTraining, Training updatedTraining) {
+            existingTraining.setTrainingName(updatedTraining.getTrainingName());
+            existingTraining.setTrainingTypeId(updatedTraining.getTrainingTypeId());
+            existingTraining.setTrainingDate(updatedTraining.getTrainingDate());
+            existingTraining.setTrainingDuration(updatedTraining.getTrainingDuration());
+            existingTraining.setTrainee(updatedTraining.getTrainee());
+            existingTraining.setTrainer(updatedTraining.getTrainer());
+    }
```
Comment (maximelyashevich) [inline src/main/java/com/epam/gym/util/UpdateEntityFields.java:41]: Please format your code.

File: src/test/java/com/epam/gym/api/TrainerControllerIntegrationTest.java [hunk 15]
```diff
@@ -40,15 +39,13 @@
...
 import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.delete;
 import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
 import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.patch;
 import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
 import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.put;
 import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
-import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
 import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
 
 @ExtendWith({SpringExtension.class, TrainerServiceParameterResolver.class})
```
Comment (maximelyashevich) [inline src/test/java/com/epam/gym/api/TrainerControllerIntegrationTest.java:51]: Can we consider `@SpringBootTest` for this class?

File: src/main/java/com/epam/gym/dto/UserCredentials.java [hunk 16]
```diff
@@ -2,6 +2,8 @@
 
 import jakarta.validation.constraints.NotBlank;
 import lombok.Builder;
+import lombok.Getter;
+import lombok.Setter;
```
Comment (maximelyashevich) [inline src/main/java/com/epam/gym/dto/UserCredentials.java:6]: Unused imports:
```suggestion
```

**Review by maximelyashevich — APPROVED** [review]
Good job! Reviewed from my side.
